### ISSUE TITLE
Improper use could cause undefined behaviour

### ISSUE THREAD

**Issue by dominik-korsa** [google/command-fds#22]
There are two places I can think of in this library where UB can occur in safe code:
## 1. `CommandFdExt::fd_mappings`
The `fd_mappings` function has the following warning in the documentation
> Warning: Calling this more than once on the same command, or attempting to run the same command more than once after calling this, may result in unexpected behaviour.

This clearly violates the rule put forward in [The Rustonomicon](https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html):
> No matter what, Safe Rust can't cause Undefined Behavior.

This means that the library should either:
- Have runtime checks to prevent calling the `fd_mappings` or `spawn` function more than once. Panicking is *not* considered unsafe\
This is analogous to the indexing (`[]`) operator of slices - trying to access an out of bounds index causes a panic
- Mark the `fd_mappings` function as `unsafe` - this makes it clear that there is a contract that needs to be fulfilled in order to call the function and the user must assert they understand it by using the `unsafe {}` block\
This is analogous to the unsafe [`get_unchecked`](https://doc.rust-lang.org/std/primitive.slice.html#method.get_unchecked) method of a slice

In both cases the documentation should be updated accordingly (replacing the warning with a `Panics` or `Safety` section)

## 2. `FdMapping`
The other potential source of UB is the `FdMapping` struct. As per the documentation for `FdMapping`:
> The parent_fd must be kept open until after the child is spawned.

This is clearly another contract that must be fulfilled by the crate user and improper usage can and will lead to UB. While getting the `RawFd` from an `OwnedFd` (or any other type - using the [`IntoRawFd`](https://doc.rust-lang.org/std/os/fd/trait.IntoRawFd.html) trait) is not unsafe, using a `RawFd` is. The logic is essentially the same as with raw pointers:
- Getting a raw pointer is safe (see [`Vec::as_ptr`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_ptr))
- Dereferencing a raw pointer is unsafe

Rust's standard library already provides a way to use file descriptors safely: the [`OwnedFd`](https://doc.rust-lang.org/std/os/fd/struct.OwnedFd.html) and [`BorrowedFd`](https://doc.rust-lang.org/std/os/fd/struct.BorrowedFd.html) types. See the [I/O safety section in std::io docs](https://doc.rust-lang.org/std/io/index.html#io-safety) for more details. They can be used to ensure that `parent_fd` will in fact be opened while we have a reference to a `OwnedFd` or `BorrowedFd`.

Fixing this problem is a bit more nuanced. The simple option would be to make either the `fd_mappings` function or `FdMapping` constructor `unsafe` (well, currently there is no `FdMapping::new` function, so the interface would need to be changed - [thread](https://internals.rust-lang.org/t/type-construction-safety-unsafe-struct-unsafe-drop/11323)) and document the safety contract (namely: the user must ensure `parent_fd` is not closed until the child is spawned). 

A more sound solution could be making it so that `fd_mappings` has a safe variant which takes either `BorrowedFd`s or `OwnedFd`s for the `parent_fd`. (I'm not sure how should `child_fd` be implemented, as it doesn't refer to an existing `fs` - leaving it as a `RawFd` might be a valid solution). 

The usage could look like this (modified example from the docs):

With `BorrowFd`:
```rs 
let file = File::open("Cargo.toml").unwrap();

let mut command = Command::new("ls");
command.arg("-l").arg("/proc/self/fd");

let fd: BorrowedFd = file.as_fd();

command
    .fd_mappings(vec![
        FdMapping::from_borrowed_fd(
            /* parent_fd: */ fd, // or &fd ?
            /* child_fd: */ 3,
        ),
    ])
    .unwrap();

// We must use lifetime annotations to make sure the borrowed fd is still valid at this point
// This might require creating a wrapper over `Command`
let mut child = command.spawn().unwrap();

// The file is still valid at this point, but two processes have file descriptors to it
// Are there any valid use cases for this?
// The behaviour could be unexpected, as two processes have access to the same file
file.write(/*...*/);

// We can drop the file - this closes the file descriptor owned by file,
// but no the duplicated one, used by the new thread
drop(file);

child.wait().unwrap();
```

With `OwnedFd`:
```rs 
let file = File::open("Cargo.toml").unwrap();

let mut command = Command::new("ls");
command.arg("-l").arg("/proc/self/fd");

let owned_fd = OwnedFd::from(file);
// OwnedFd::from<File> consumes `file`
// so the variable `file` is no longer valid at this point

command
    .fd_mappings(vec![
        FdMapping::from_owned_fd(
            /* parent_fd: */ owned_fd, // We transfer ownership of owned_fd
            /* child_fd: */ 3,
        ),
    ])
    .unwrap();

let mut child = command.spawn().unwrap();
// OwnedFd (just as File) implements Drop, which closes the owned stream
// The internal implementation should consider when the OwnedFd is dropped
// Of course it should still be valid when we make the `dup2` syscall
// We should probably examine how it is implemented for standard streams in std::process
// 
// The documentation should make it clear when the original file descriptor will be closed
// This is especially important when using pipes, where we expect all the descriptors on the writeable end to be closed,
// so that the reading end can end with an EOF

child.wait().unwrap();
```

**Comment by qwandor** [comment]
I think `FdMapping.parent_fd` would ideally be a `BorrowedFd`, and `CommandFdExt::preserved_fds` should also take `BorrowedFd`s. The problem is that they should have a lifetime that lasts at least as long as the `Command`, but there is no way of expressing this. Requiring them to have a `'static` lifetime would be correct but not useful, as most `BorrowedFd`s are constructed from an `OwnedFd` with a finite lifetime.

Taking `OwnedFd` instead would avoid the lifetime issues, but may require the caller to duplicate the file descriptor first, which is not ideal.

**Comment by dominik-korsa** [comment]
> On further reflection I think `OwnedFd` is reasonable, as it's unlikely that the parent process would want to use the file descriptors further after spawning a child with them, and if it really wants to it can dup them. Please take a look at #23 and #25.

I also think taking `OwnedFd` is the right solution in this case, due to the lack of a lifetime annotation on the `Command` struct. I would just suggest documenting when the `OwnedFd` is dropped (and so when is the file descriptor closed).

Consider this example, which uses the [`os_pipe`](https://crates.io/crates/os_pipe) crate:
```rust
fn get_fd3_output() -> Vec<u8> {
    let (mut reader, writer) = os_pipe::pipe().unwrap();
    
    // Let's assume we have a "my_bin" executable which doesn't take any inputs
    // and it "prints" it's output to FD3.
    // Let's also assume the output is very long and will not fit in the pipe buffer.
    let command = Command::new("my_bin")
        .fd_mappings(vec![FdMapping {
            parent_fd: writer.into(),
            child_fd: 3
        }]).expect("FD mappings should be valid")
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .stdin(Stdio::null());
    
    let mut child = command.spawn().unwrap();
    
    let mut buf: Vec<u8> = vec![];
    // The .read_to_end method will block until it receives an EOF
    // But we will not see EOF until all FDs to the writeable end of the pipe are closed
    // If the OwnedFd is not dropped at this point, a deadlock will occur!
    reader.read_to_end(&mut buf);
    
    child.wait().unwrap();
    buf
}
```

Would the behaviour be any different if we were to do `drop(command)` earlier?
```rust
fn get_fd3_output_v2() -> Vec<u8> {
    let (mut reader, writer) = os_pipe::pipe().unwrap();

    let command = Command::new("my_bin")
        .fd_mappings(vec![FdMapping {
            parent_fd: writer.into(),
            child_fd: 3
        }]).expect("FD mappings should be valid")
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .stdin(Stdio::null());
    
    let mut child = command.spawn().unwrap();
    drop(command); // <-- will this drop the inner OwnedFd?
    
    let mut buf: Vec<u8> = vec![];
    // If the OwnedFd is not dropped at this point, a deadlock will occur!
    reader.read_to_end(&mut buf);
    
    child.wait().unwrap();
    buf
}
```

**Comment by qwandor** [comment]
Good point. Added documentation of this to #25.

**Comment by qwandor** [comment]
Fixed by #23 and #25.
